src/trajectory.jl

```diff
@@ -218,190 +282,167 @@ Base.show(io::IO, d::Termination) = print(io, "Termination(dynamic=$(d.dynamic),
+    θ0minusθ1 = z0.θ - z1.θ
+    s = (dot(-θ0minusθ1, ∂H∂r(h, -z0.r)) >= 0) || (dot(θ0minusθ1, ∂H∂r(h, z1.r)) >= 0)
+    return Termination(s, false)
 end
 
 """
-Sample a condidate point form two trees (`tleft` and `tright`) under multinomial sampling.
+Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
+using the generalised no-U-turn criterion.
+
+Ref: https://arxiv.org/abs/1701.02434
 """
-function combine(
-    rng::AbstractRNG,
-    tleft::FullBinaryTree{MultinomialTreeSampler{F}},
-    tright::FullBinaryTree{MultinomialTreeSampler{F}}
-) where {F<:AbstractFloat}
-    return rand(rng) < exp(tleft.sampler.logw - logaddexp(tleft.sampler.logw, tright.sampler.logw)) ? tleft.zcand : tright.zcand
+function isterminated(h::Hamiltonian, t::FullBinaryTree{C}) where {S,C<:GeneralisedNoUTurn}
+    # z0 is starting point and z1 is ending point
+    z0, z1 = t.zleft, t.zright
+    rho = t.c.rho
+    s = (dot(rho, ∂H∂r(h, -z0.r)) >= 0) || (dot(-rho, ∂H∂r(h, z1.r)) >= 0)
+    return Termination(s, false)
 end
 
 """
 Recursivly build a tree for a given depth `j`.
 """
 function build_tree(
     rng::AbstractRNG,
-    nt::NUTS{I,F,S},
+    nt::NUTS{S,C,I,F},
     h::Hamiltonian,
     z::PhasePoint,
     sampler::AbstractTreeSampler,
     v::Int,
     j::Int,
     H0::AbstractFloat
-) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractTreeSampler}
+) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractTreeSampler,C<:AbstractTerminationCriterion}
     if j == 0
         # Base case - take one leapfrog step in the direction v.
         z′ = step(nt.integrator, h, z, v)
-        H′ = -neg_energy(z′)
-        basesampler = makebase(sampler, H′)
-        termination = isterminated(basesampler, nt, H0, H′)
+        H′ = energy(z′)
         α′ = exp(min(0, H0 - H′))
-        return FullBinaryTree(z′, z′, z′, basesampler, termination, α′, 1)
+        sampler′ = S(sampler, H0, z′)
+        return FullBinaryTree(z′, z′, C(z′), α′, 1), sampler′, Termination(sampler′, nt, H0, H′)
     else
         # Recursion - build the left and right subtrees.
-        t′ = build_tree(rng, nt, h, z, sampler, v, j - 1, H0)
+        t′, sampler′, termination′ = build_tree(rng, nt, h, z, sampler, v, j - 1, H0)
         # Expand tree if not terminated
-        if !isterminated(t′.termination)
+        if !isterminated(termination′)
             # Expand left
             if v == -1
-                t′′ = build_tree(rng, nt, h, t′.zleft, sampler, v, j - 1, H0) # left tree
+                t′′, sampler′′, termination′′ = build_tree(rng, nt, h, t′.zleft, sampler, v, j - 1, H0) # left tree
                 tleft, tright = t′′, t′
             # Expand right
             else
-                t′′ = build_tree(rng, nt, h, t′.zright, sampler, v, j - 1, H0) # right tree
+                t′′, sampler′′, termination′′ = build_tree(rng, nt, h, t′.zright, sampler, v, j - 1, H0) # right tree
                 tleft, tright = t′, t′′
```

Comment:
Same here: `tleft` and `tright` ==> `trajetory_left` and `trajectory_right` respectively.

Reply:
Would do `tree_left` and `tree_right`.